src/lib/common/include/sol-reentrant.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file
+ * @brief Structure and macros for making structures reentrancy-proof
+ */
+
+/**
+ * @defgroup Reentrant Reentrant
+ *
+ * @{
+ */
+
+/**
+ * @struct sol_reentrant
+ *
+ * @brief Structure containing the flags for safely freeing a larger structure
+ *
+ * This structure is meant to be used as the first field of larger structures
+ * which are affected by calls to external callbacks that in turn end up
+ * calling library APIs. The possibility of a double free is particularly
+ * likely in such cases.
+ */
+struct sol_reentrant {
+    /**
+     * @brief Structure is in use
+     */
+    bool in_use;
+    /**
+     * @brief Structure is stale and should be freed as soon as possible
+     */
+    bool is_stale;
+};
+
+/**
+ * @brief Perform a function call to an external callback
+ *
+ * @param reentrant The reentrant affected by the function call
+ *
+ * @param call_statement The function call statement
+ *
+ * Wraps a call to a function by first recording whether the reentrant
+ * structure was in use and, after the function call, restoring the state of
+ * the reentrant structure from the recorded value. Unlike a monitor, which
+ * protects the list of callbacks itself, this wrapper protects the library's
+ * structure from being deleted as part of the external callback, if it's still
+ * in use after the callback completes.
+ */
+#define SOL_REENTRANT_CALL(reentrant, call_statement) \
+    do { \
+        struct sol_reentrant *local = ((struct sol_reentrant *)(reentrant)); \
```

Comment:
Also, unlike a monitor, the context is not one provided along with the callback. The context is a structure internal to soletta which the callback may destroy by calling a soletta API. The place where I came across this was (in a hybrid JS/C pseudocode):
```
pending = sol_oic_client_find_resources( function( resource ) {
  sol_oic_pending_cancel( pending );
  return NULL;
} );
```

So, the context was being destroyed twice: Once during the action of the callback because it was explicitly freeing the context via ```sol_oic_pending_cancel()```, and again in response to the NULL return value, indicating that the context is no longer necessary. The pending is an internal Soletta structure, rather than one specified along with the discovery response callback.